PHP 8.0 and 8.1

Tybaze opened the pull request:
# PHP 8.0 & 8.1 compatibility

[No Enhancement - please considere to use Doctrine2 for your new Project]

Target : Maintains maximum backward compatibility down to 5.3 for the moment

PHP 8.0 and 8.1 brings a lot of incompatibility.
https://www.php.net/manual/en/migration80.incompatible.php
https://www.php.net/manual/en/migration81.incompatible.php

## Tests

I have to update travis, to include new php versions, and to remove php5.3 on Precise (letsencrypt root certificate is broken, impossible to download composer anymore)

I tested localy on PHP 5.3 and passed the build successfully.

![alt text](https://api.travis-ci.com/Tybaze/doctrine1.svg?branch=compat_php8.1)
[Travis Build Link](https://api.travis-ci.com/Tybaze/doctrine1.svg?branch=compat_php8.1)

| PHP version  |  Status |
| ------------- | --- |
| php5.3  |  🟢 |
| php5.4  |  🟢 |
| php5.5  |  🟢 |
| php5.6  |  🟢 |
| php7.0  |  🟢 |
| php7.1  |  🟢 |
| php7.2  |  🟢 |
| php7.3  |  🟢 |
| php7.4  |  🟢 |
| php8.0  |  🟢 |
| php8.1  |  🟢 |

## Future

One Day PHP 9 will broke compatibility with php < 7.4 due to the lacks of "mixed" keywords implementations.

Please considere rewriting asap your project to another framework, like Doctrine2 if your app cannot not support at least PHP 7.4 within a few months/years

alquerci commented:
Hello @Tybaze 

Thanks for your contribution for the huge goal of PHP 8 support.

Did you execute tests based on the fix of tests execution, see #56 ?

Tybaze commented:
> Did you execute tests based on the fix of tests execution, see #56 ?

No sorry didn't see it, I run test directly with travis. 
I'll try them and #86 on monday.

Tybaze commented:
I was using a wrong way to run test  this PR 

I have reviewed #56 and patched my branch localy to do the test.
I also use your new #86 with docker compose, great thanks @alquerci 

Now I have updated the first post of the PR, new status will be, once #56 will be merged

| PHP version  |  Status |
| ------------- | --- |
| php8.0  |  🟢 |
| php8.1  |  🔴 |

Working on PHP8.1...

Tybaze commented:
Everything is working now on PHP8.0 and PHP8.1 👍 

How do I test : 
```
git clone git@github.com:Tybaze/doctrine1.git
git remote add alquerci https://github.com/alquerci/doctrine1.git
git fetch alquerci

# PR85
git checkout compat_php8.1

# PR86 
git merge alquerci/add-docker-compose

# PR56 
git merge alquerci/fix-tests-to-be-able-to-finish-it-without-a-fatal-error

# Fix a very small conflict in .gitignore
git commit 

# modify file ./tests/bin/test , replace variable "skipPHPVersions" with "php9"
# Then run the test
./tests/bin/test
```
And we get it :
| PHP version  |  Status |
| ------------- | --- |
| php5.3  |  🟢 |
| php5.4  |  🟢 |
| php5.5  |  🟢 |
| php5.6  |  🟢 |
| php7.0  |  🟢 |
| php7.1  |  🟢 |
| php7.2  |  🟢 |
| php7.3  |  🟢 |
| php7.4  |  🟢 |
| php8.0  |  🟢 |
| php8.1  |  🟢 |

thePanz commented on lib/Doctrine/Connection/Profiler.php:
```diff
@@ -109,7 +109,7 @@ public function __call($m, $a)
      * @param mixed $key
      * @return Doctrine_Event
```
```suggestion
     * @return Doctrine_Event|null
```

thePanz commented on lib/Doctrine/Connection/Profiler.php:
```diff
@@ -123,7 +123,7 @@ public function get($key)
      *
      * @return array        all events in an array
```
```suggestion
     * @return Doctrine_Event[] All events in an array
```

thePanz commented on lib/Doctrine/Connection/Profiler.php:
```diff
@@ -154,7 +156,7 @@ public function count()
      *
      * @return Doctrine_Event
```
```suggestion
     * @return Doctrine_Event|null
```

thePanz commented on lib/Doctrine/Connection/Sqlite.php:
```diff
@@ -87,8 +93,18 @@ public function connect()
             return false;
```
The method is annotated to return void, let's not return a boolean :)
```suggestion
            return;
```

alquerci replied:
Sadly, it is the PHP Doc that is wrong here.

The interface to respect is defined on [Doctrine_Connection::connect()](https://github.com/FriendsOfSymfony1/doctrine1/blob/master/lib/Doctrine/Connection.php#L455)

Tybaze replied:
908bcbb316aa372ed887a4973121af5b8d890ee7 I have set a "@return boolean", exactly like the abstract Doctrine_Connection::connect()

is it ok ?

thirsch replied:
Hi @thePanz, you've given a thumbs up to the change from @Tybaze therefore it should be ok to accept this pr, shouldn't it?

thePanz commented on lib/Doctrine/Record.php:
```diff
@@ -798,6 +798,33 @@ private function prepareIdentifiers($exists = true)
+        $array = unserialize($serialized);
+
+        $this->__unserialize($array);
+
```
Remove the extra line here

Tybaze replied:
Fixed !

thePanz commented on lib/Doctrine/Table.php:
```diff
@@ -2991,26 +3029,30 @@ public function serialize()
+        $this->_useIdentityMap = $data[10];
     }
 
+
```
Remove the extra line here

Tybaze replied:
Fixed, and also reviewed phpdoc which trigger some errors in my IDE.
Thank you !

thePanz commented on tests/Ticket/1783TestCase.php:
```diff
@@ -12,9 +12,12 @@ public function testValidateLargeIntegers()
-        
-        $this->assertTrue($test->isValid());
+
+        // This test works on php 32bit version because float allow to represent bigger value than a int.
```
Good catch! :+1:

thirsch commented on lib/Doctrine/Table.php:
```diff
@@ -1132,6 +1132,11 @@ public function processOrderBy($alias, $orderBy, $columnNames = false)
         }
 
+        // Php8.1 require a string
+        if(null === $orderBy) {
```
```suggestion
        if (null === $orderBy) {
```

thirsch commented on lib/Doctrine/Task.php:
```diff
@@ -53,7 +53,7 @@ public function __construct($dispatcher = null)
         $this->dispatcher = $dispatcher;
 
-        $taskName = $this->getTaskName();
+        $taskName = (string)$this->getTaskName();
```
```suggestion
        $taskName = (string) $this->getTaskName();
```

thirsch commented on lib/Doctrine/Validator.php:
```diff
@@ -126,9 +126,9 @@ public static function validateLength($value, $type, $maximumLength)
     {
         if (function_exists('mb_strlen')) {
-            return mb_strlen($string, 'utf8');
+            return mb_strlen((string)$string, 'utf8');
```
```suggestion
            return mb_strlen((string) $string, 'utf8');
```

thirsch commented on lib/Doctrine/Validator.php:
```diff
@@ -126,9 +126,9 @@ public static function validateLength($value, $type, $maximumLength)
+            return mb_strlen((string)$string, 'utf8');
         } else {
-            return strlen(utf8_decode($string));
+            return strlen(utf8_decode((string)$string));
```
```suggestion
            return strlen(utf8_decode((string) $string));
```

thirsch commented on lib/Doctrine/Validator/Notblank.php:
```diff
@@ -41,6 +41,6 @@ class Doctrine_Validator_Notblank extends Doctrine_Validator_Driver
     public function validate($value)
     {
-        return (trim($value) !== '' && $value !== null);
+        return (trim((string) $value) !== '' && $value !== null);
```
You could either swap the two operands and remove the cast or remove the second condition.

Tybaze replied:
Thank you very much, it avoid a useless cast.

thePanz commented on lib/Doctrine/Table.php:
```diff
@@ -2991,26 +3027,35 @@ public function serialize()
+    /**
+     * Table loaded from cache, instead of __construct()
+     *
+     * @param Doctrine_Connection $conn
```
Removing, as it does not bring more clarity :)

```suggestion
```

alquerci commented on lib/Doctrine/Connection/Sqlite.php:
```diff
@@ -84,11 +90,21 @@ public function __construct(Doctrine_Manager $manager, $adapter)
     {
         if ($this->isConnected) {
-            return false;
+            return;
```
Needs to respect the interface defined on [Doctrine_Connection::connect()](https://github.com/FriendsOfSymfony1/doctrine1/blob/master/lib/Doctrine/Connection.php#L455)

The PHP Doc is wrong here.

alquerci commented on tests/Ticket/1783TestCase.php:
```diff
@@ -12,9 +12,12 @@ public function testValidateLargeIntegers()
+
+        // This test works on php 32bit version because float allow to represent bigger value than a int.
+        // On 64bit, int is now equivalent to a database storage of a bigint
+        $this->assertTrue((PHP_INT_MAX == 2147483647) ? $test->isValid() : true);
```
Yes, nice catch, it is a good track, but the conclusion is wrong.

The table definition will correspond to MySql type `BIGINT UNSIGNED`.

Which has a maximum value of `2^64 - 1`.
However, `PHP_INT_MAX` on 64bit system architecture is `2^63-1` as PHP integer is not unsigned.

Using `getErrorStackAsString()` we can see what validator failed.
We can see that the error comes from `unsigned` validation.

* More over, there is no validation on the length for integer.
    https://github.com/FriendsOfSymfony1/doctrine1/blob/master/lib/Doctrine/Validator.php#L95-L96
* Digging deeper of the `Unsigned` validator, you can see that the regex is in fault.
    It depends on how the value is converted to string, when it is bigger than PHP_INT_MAX it contains `E` the exponential symbol.

**Conclusion**
The issue still exists for 64bit system.

For now, we can ignore this test.
https://github.com/FriendsOfSymfony1/doctrine1/pull/56/commits/6a08c161467bddf89afd0eeeacfa6abf085d2c47

alquerci commented on lib/Doctrine/Task.php:
```diff
@@ -53,7 +53,7 @@ public function __construct($dispatcher = null)
         $this->dispatcher = $dispatcher;
 
-        $taskName = $this->getTaskName();
+        $taskName = (string) $this->getTaskName();
```
There is no test case when task name is an empty string.

So the minimum things to do should be
* Handle `null` as @thePanz said
* Add test case for empty string, it seems an extreme edge case, but why not,

Tybaze replied:
Fixed by handling the null value : 711ec88b83e0a9ab64d1c7765206e220cbfca85a

I miss some time to create the unit test about this rare case

alquerci replied:
> I miss some time to create the unit test about this rare case

How it is or will be going ?

This try to add a proof that it work as it should, is like important for the ability to maintain doctrine1 code.

If you have time, I will be interested in your return experience.

---

Between I have a doubt on the last commit https://github.com/FriendsOfSymfony1/doctrine1/commit/86519ea755905c2004932fd3b40b321e1d0543e2 behaviour.

alquerci replied:
@thirsch It left this issue.

Currently, custom task name is ignored when set on `taskName` property.

```php
if (null === $taskName || 0 !== strlen($taskName)) {
```

Should be at least

```php
if (null === $taskName || 0 === strlen($taskName)) {
```

Tybaze replied:
Just fixed in 09da1cd899410b2a026fae178ec57440baa62e2a

thePanz replied:
@Tybaze if we cast to string `(string) $this->getTaskName();` , then the `if (null === $taskName` will always be false.

With the casting we can simplify the condition to

```php
$taskName = (string) $this->getTaskName();
if ('' === trim($taskName)) {
    $taskName = self::deriveTaskName(get_class($this));
}
```

What do you think?

thePanz replied:
This covers the weird case of a task with a set of spaces as name

alquerci replied:
@thePanz you are right.

There are many ways to make this work.

But the main issue here is that there no test for that case.

> How can you prevent harm to the structure of your code if you don’t have the tests that allow you to clean it?
--- "Clean Craftsmanship: Disciplines, Standards, and Ethics (Robert C. Martin Series) (English Edition)" write by Robert C. Martin

---

Let me try to add a repeatable proof that every element of the code works as it should.

alquerci replied:
Here we go

https://github.com/Tybaze/doctrine1/pull/2

alquerci commented on lib/Doctrine/Connection/Statement.php:
```diff
@@ -317,7 +317,7 @@ public function execute($params = array())
     public function fetch($fetchMode = Doctrine_Core::FETCH_BOTH,
                           $cursorOrientation = Doctrine_Core::FETCH_ORI_NEXT,
-                          $cursorOffset = null)
+                          $cursorOffset = 0)
```
As it is, `null` on the interface, this is a violation.

Can you try to fix without this violation?

alquerci commented on lib/Doctrine/Table.php:
```diff
@@ -2991,26 +3027,35 @@ public function serialize()
     }
 
+    /**
+     * Table loaded from cache, instead of __construct()
```
The method name is a bit misleading.

What does it do ?

```suggestion
     * Creates new instance and initialize it from cache.
```

alquerci commented on lib/Doctrine/Table.php:
```diff
@@ -2991,26 +3027,35 @@ public function serialize()
+     * Table loaded from cache, instead of __construct()
+     *
+     * @param Doctrine_Connection $conn
+     * @return void
```
```suggestion
```

alquerci commented on lib/Doctrine/Table.php:
```diff
@@ -2991,26 +3027,35 @@ public function serialize()
+    /**
+     * Unserializes a Doctrine_Record instance for php 7.4+
+     *
+     * @param array $serialized
```
```suggestion
     * @param array $data
```

alquerci commented:
A pass of code structure fixes will be appreciated.

Tybaze commented on lib/Doctrine/Connection/Sqlite.php:
```diff
@@ -84,11 +90,21 @@ public function __construct(Doctrine_Manager $manager, $adapter)
+            // PHP8.1 require default to true to keep BC
+            // https://www.php.net/manual/en/migration81.incompatible.php#migration81.incompatible.pdo.sqlite
+            // Can be overwritten by user later
+            $this->dbh->setAttribute(PDO::ATTR_STRINGIFY_FETCHES, true);
```
In the case if the user modify dbh attribute with PDO::ATTR_STRINGIFY_FETCHES, he do not respect doctrine1 behavior, it will be overwritten.

If the user set Doctrine_Core::ATTR_STRINGIFY_FETCHES to false, the value will still be false, and the new php 8.1 behavior will apply.

Doctrine3 decided to drop the handling, delegating the fault to the pdo driver. This is normal for an evolutive framework to move to the new behavior
https://github.com/doctrine/dbal/pull/5229/files

As doctrine1 target is to keep max BC, it's better to do our max to keep the previous behavior

thePanz commented on lib/Doctrine/Task.php:
```diff
@@ -56,7 +56,7 @@ public function __construct($dispatcher = null)
 
         //Derive the task name only if it wasn't entered at design-time
-        if (! strlen($taskName)) {
+        if (null === $taskName || ! strlen($taskName)) {
```
```suggestion
        if (null === $taskName || 0 !== strlen($taskName)) {
```

alquerci replied:
@thePanz I do not understand this suggestion. It changes the behaviour or the structure ?

thePanz replied:
imho it makes more explicit the strlen evaluation, and the fact that strlen returns an integer (not a boolean)

Tybaze replied:
Fixed on 09da1cd899410b2a026fae178ec57440baa62e2a

Tybaze commented:
Hi Everyone ^^
Do I miss something to merge this PR ?

alquerci commented:
@thePanz I remind you that this PR required following ones

* FriendsOfSymfony1/doctrine1#56
* FriendsOfSymfony1/doctrine1#86

Tybaze commented:
Rebase:
Merge commits about the TaskName declaration, using the latest @thePanz version:
`if ('' === trim($taskName)) {`

Merge commits review of this PR about annotation, and conding style.

thePanz commented:
@Tybaze coudl you rebase on the latest `master` branch?
- the first two commits from xNatek are already merged
- squash commits to group related changes togerter
- remove the "merge master" commits from my side (I tried to use the GitHub feature, but we should not have those commits in the project's history)

Thanks! then we will be able to merge this! :tada:

Tybaze commented:
> @Tybaze coudl you rebase on the latest `master` branch?
> 
>     * the first two commits from xNatek are already merged
> 
>     * squash commits to group related changes togerter
> 
>     * remove the "merge master" commits from my side (I tried to use the GitHub feature, but we should not have those commits in the project's history)
> 
> 
> Thanks! then we will be able to merge this! 🎉

- Rebased on master (including xNatek commit)
- Squashed by modification type (Compat 8.0, 8.1, PR Review)
- Cleaned Merge master

@thePanz  You're welcome ! 
Sorry for the delay ^^

thePanz commented:
Merged @Tybaze @alquerci 
Thanks for your efforts!
